Simplify uefi::Result type and remove Completion

nicholasbishop opened the pull request:
* Change `uefi::Result::Ok` variant to not use `Completion`.
* Drop `Completion` type entirely.
* Drop `ResultExt`'s `unwrap_success`, `expect_success`, `expect_error`,
  and `map_inner` methods. The standard `Result` methods can be used
  instead.
* Add `ResultExt::handle_warning`.
* Drop `ResultExt::log_warning`, `handle_warning` can be used instead if
  needed.
* Added `Output::output_string_lossy` which ignores `WARN_UNKNOWN_GLYPH`
  warnings.
* Updated `set_variable` doc to describe a warning it can return.

Implements https://github.com/rust-osdev/uefi-rs/issues/360

GabrielMajeri commented:
Thank you for coming with an initial implementation of the RFC. I'd say we keep this open for a few days to gather feedback then proceed with the change, as mentioned in https://github.com/rust-osdev/uefi-rs/issues/360#issuecomment-1043442037.

josephlr commented on src/lib.rs:
```diff
@@ -51,7 +51,7 @@ pub use self::data_types::{unsafe_guid, Identify};
 
 mod result;
-pub use self::result::{Completion, Result, ResultExt, Status};
+pub use self::result::{Result, ResultExt, Status};
```
As mentioned in https://github.com/rust-osdev/uefi-rs/issues/360#issuecomment-1057738309 should we also be exporting `Error` here?

nicholasbishop commented:
Rebased to pull in the change that makes `Error` public, also added `ResultExt::handle_warning` as suggested in https://github.com/rust-osdev/uefi-rs/issues/360#issuecomment-1057736468.

josephlr commented on src/result/mod.rs:
```diff
@@ -26,81 +21,51 @@ mod status;
 pub use self::status::Status;
 
 /// Return type of most UEFI functions. Both success and error payloads are optional.
```
Should we mention here something about how warnings are treated as errors?

nicholasbishop replied:
Good call, I've overhauled the documentation here a bit. The module wasn't public so I dropped most of the module docstring and expanded the `Result` docstring instead, since that's probably where it's most useful anyway.

josephlr commented on src/result/mod.rs:
```diff
@@ -26,81 +21,51 @@ mod status;
 
-    /// Treat warnings as errors
-    fn warning_as_error(self) -> core::result::Result<Output, Error<ErrData>>
+    /// Calls `op` if the result contains a warning, otherwise returns
+    /// the result unchanged.
```
It seems like this method should:
  - Say what the default behavior is (i.e. that warnings are treated as errors)
  - Mention that users shouldn't _normally_ call this function (i.e. it's reserved for special cases)
  - Mention the similarities to [`Result::or_else`](https://doc.rust-lang.org/nightly/std/result/enum.Result.html#method.or_else)

Do we also want to have a code example for this function somewhere? (I'm not sure what the rest of this crate does).

nicholasbishop replied:
Updated the docstring to include those things and an example.

josephlr reviewed (approved):
Seems reasonable to me!

nicholasbishop commented:
Updated with the latest `main` changes, also added changelog entries.

@GabrielMajeri if it looks good to you I'd like to move forward with merging this since the feedback in https://github.com/rust-osdev/uefi-rs/issues/360 has been positive.

GabrielMajeri commented:
@nicholasbishop Agreed, everything looks fine, let's merge this!
